src/cli.rs

```diff
@@ -204,4 +213,77 @@ mod tests {
         assert!(parse_args(&["supply-chain", "supply-chain", "json", "--print-schema"]).is_err());
         assert!(parse_args(&["supply-chain", "supply-chain", "crates", "-d"]).is_err());
     }
+
+    #[test]
+    fn ui() {
+        update();
+
+        for entry in read_dir("ui").unwrap() {
+            let entry = entry.unwrap();
+            let path = entry.path();
+
+            if path.extension() != Some(OsStr::new("version")) {
+                continue;
+            }
+
+            let name = path.file_stem().and_then(OsStr::to_str).unwrap();
+
+            let contents = read_to_string(&path).unwrap();
+            let version = contents.trim_end();
+
+            let tempdir = download(name, version);
+
+            for subcommand in ["crates", "publishers"] {
+                for no_dev in [false, true] {
+                    let path = path.with_extension(
+                        String::from(subcommand) + if no_dev { "_no_dev" } else { "" },
+                    );
+                    let stdout_expected = read_to_string(&path).unwrap();
+
+                    let mut command = Command::new(
+                        Path::new(env!("CARGO_MANIFEST_DIR"))
+                            .join("target/debug/cargo-supply-chain"),
+                    );
+                    command.current_dir(tempdir.path().join(format!("{name}-{version}")));
+                    command.args(["supply-chain", subcommand]);
+                    if no_dev {
+                        command.arg("--no-dev");
+                    }
+                    let output = command.output().unwrap();
+                    let stdout_actual = std::str::from_utf8(&output.stdout).unwrap();
+
+                    if enabled("BLESS") {
+                        write(path, stdout_actual).unwrap();
+                    } else {
+                        assert_eq!(stdout_expected, stdout_actual);
```

Comment:
I fear this will break a lot due to Cargo changing its dependency tree. I'm not looking forward to constantly eyeballing the changes and re-running it with "BLESS" to keep the tests working.

Instead of end-to-end changes, how about we just test the dev-dependency filtering algorithm? `cargo supply-chain` gets the dependency tree from `cargo metadata`, which outputs JSON. We can make a bunch of test crates in-tree and save this JSON (or even save the `Cargo.lock` + `Cargo.toml` and run `cargo metadata` on it) and have tests that we fully control, without relying on `cargo` not to change its dependency tree.

This can be done as unit tests instead of end-to-end tests too, so this would work offline. We'd miss out on the test coverage for the UI, but it's not a particularly complex part to begin with so I'm not really worried.

Reply:
This all sounds reasonable to me.

Running `cargo metatdata` could need the network (right?). That suggests to me the JSON idea might work better.

Having said that, would it make sense to move the majority of `sourced_dependencies` (roughly [this](https://github.com/rust-secure-code/cargo-supply-chain/blob/7773a73ef4af2943b953aa3b0688c9acd7ef8a74/src/common.rs#L57-L97)) into a new function called something like `sourced_dependencies_from_metadata`, and then the tests call that new function?

Reply:
Yeah, sure, feel free to split up the code as appropriate.